flytekit/common/translator.py

```diff
@@ -0,0 +1,260 @@
+from typing import Dict, List, Optional, Union
+
+from flytekit.annotated.base_task import PythonTask, TaskMetadata
+from flytekit.annotated.condition import BranchNode
+from flytekit.annotated.container_task import ContainerTask
+from flytekit.annotated.context_manager import RegistrationSettings
+from flytekit.annotated.launch_plan import LaunchPlan, ReferenceLaunchPlan
+from flytekit.annotated.node import Node
+from flytekit.annotated.reference_task import ReferenceTask
+from flytekit.annotated.workflow import ReferenceWorkflow, Workflow, WorkflowFailurePolicy, WorkflowMetadata
+from flytekit.common import constants as _common_constants
+from flytekit.common.interface import TypedInterface
+from flytekit.common.launch_plan import SdkLaunchPlan
+from flytekit.common.nodes import SdkNode
+from flytekit.common.tasks.task import SdkTask
+from flytekit.common.workflow import SdkWorkflow
+from flytekit.models import common as _common_models
+from flytekit.models import interface as interface_models
+from flytekit.models import launch_plan as _launch_plan_models
+from flytekit.models import literals as literal_models
+from flytekit.models.common import RawOutputDataConfig
+from flytekit.models.core import identifier as _identifier_model
+from flytekit.models.core import workflow as _core_wf
+from flytekit.models.core import workflow as workflow_model
+from flytekit.models.core.workflow import BranchNode as BranchNodeModel
+
+FlyteLocalEntity = Union[
+    PythonTask, BranchNode, Node, LaunchPlan, Workflow, ReferenceWorkflow, ReferenceTask, ReferenceLaunchPlan
+]
+FlyteControlPlaneEntity = Union[SdkTask, SdkLaunchPlan, SdkWorkflow, SdkNode, BranchNodeModel]
+
+
+def to_registerable_case(settings: RegistrationSettings, c: _core_wf.IfBlock) -> _core_wf.IfBlock:
+    if c is None:
+        raise ValueError("Cannot convert none cases to registrable")
+    then_node = get_serializable(settings, c.then_node)
+    return _core_wf.IfBlock(condition=c.condition, then_node=then_node)
+
+
+def to_registrable_cases(
+    settings: RegistrationSettings, cases: List[_core_wf.IfBlock]
+) -> Optional[List[_core_wf.IfBlock]]:
+    if cases is None:
+        return None
+    ret_cases = []
+    for c in cases:
+        ret_cases.append(to_registerable_case(settings, c))
+    return ret_cases
+
+
+# Can make this part of the function in the future.
+GLOBAL_CACHE: Dict[FlyteLocalEntity, FlyteControlPlaneEntity] = {}
+
+
+def get_serializable(
+    settings: RegistrationSettings, entity: FlyteLocalEntity, fast: Optional[bool] = False
+) -> FlyteControlPlaneEntity:
+    if entity in GLOBAL_CACHE:
+        return GLOBAL_CACHE[entity]
+
+    cp_entity = None
+    # TODO: flytekit.annotated.container_task.ContainerTask needs to be excepted
+    if isinstance(entity, ReferenceTask):
```

Comment:
my preference each if condition is modeled using a hash-map and we just have functions for each transform. Thus we follow a purely functional approach

Reply:
e.g
map[Type, Callable]?

Reply:
even if we dont do the decomposition by map - i would like us to break this function into smaller functions for ease of testing and easier logic to follow

Reply:
+1 for separate functions!